From c01e1433a23f85e0ada0f899c4c79b5a9a823cff Mon Sep 17 00:00:00 2001
From: Marek 'marx' Grac <mgrac@redhat.com>
Date: Tue, 24 Mar 2015 09:18:26 +0100
Subject: [PATCH] fencing: Add support for --tls1.0

HP iLO2 firmware 2.27 has broken implementation of TLS and SSLv3 is disabled by default.
gnutls (3.4.x) has support to disable proper negotiation and use only TLS1.0 that works well.

Option --tls1.0 (tls1.0 on stdin) was added to enable this feature and fence_ilo(2) works
correctly also with this firmware.

This version differs from master branch because we use gnutls in older version.
---
 fence/agents/ilo/fence_ilo.py      |  2 +-
 fence/agents/lib/fencing.py.py     | 18 ++++++++++++++++--
 tests/data/metadata/fence_ilo.xml  |  7 ++++++-
 tests/data/metadata/fence_ilo2.xml |  7 ++++++-
 4 files changed, 29 insertions(+), 5 deletions(-)

diff --git a/fence/agents/ilo/fence_ilo.py b/fence/agents/ilo/fence_ilo.py
index 130f27e..b9be62a 100644
--- a/fence/agents/ilo/fence_ilo.py
+++ b/fence/agents/ilo/fence_ilo.py
@@ -65,7 +65,7 @@ def define_new_opts():
 		"order" : 1}
 
 def main():
-	device_opt = ["ipaddr", "login", "passwd", "ssl", "notls", "ribcl", "web"]
+	device_opt = ["ipaddr", "login", "passwd", "ssl", "notls", "tls1.0", "ribcl", "web"]
 
 	atexit.register(atexit_handler)
 
diff --git a/fence/agents/lib/fencing.py.py b/fence/agents/lib/fencing.py.py
index 559bd24..8bafcce 100644
--- a/fence/agents/lib/fencing.py.py
+++ b/fence/agents/lib/fencing.py.py
@@ -208,7 +208,19 @@ all_opt = {
 				"This should only be used for devices that do not support TLS1.0 and up.",
 		"default" : "1",
 		"required" : "0",
-		"shortdesc" : "Disable TLS negotiation",
+		"shortdesc" : "Disable TLS negotiation, force SSL 3.0",
+		"order" : 1},
+	"tls1.0" : {
+		"getopt" : "",
+		"longopt" : "tls1.0",
+		"help" : "--tls1.0                       "
+				"Disable TLS negotiation and force TLS1.0\n"
+				"                                        "
+				"This should only be used for devices that\n"
+				"                                        "
+				"do not support TLS1.1 and up.",
+		"required" : "0",
+		"shortdesc" : "Disable TLS negotiaton, force TLS 1.0",
 		"order" : 1},
 	"port" : {
 		"getopt" : "n:",
@@ -972,7 +984,9 @@ def _open_ssl_connection(options):
 	gnutls_opts = ""
 	ssl_opts = ""
 
-	if options.has_key("--notls"):
+	if options.has_key("--tls1.0"):
+		gnutls_opts = "--priority \"NORMAL:-VERS-TLS1.2:-VERS-TLS1.1:+VERS-TLS1.0:-VERS-SSL3.0\""
+	elif options.has_key("--notls"):
 		gnutls_opts = "--priority \"NORMAL:-VERS-TLS1.2:-VERS-TLS1.1:-VERS-TLS1.0:+VERS-SSL3.0\""
 
 	# --ssl is same as the --ssl-insecure; it means we DO NOT want to verify certificate in these cases
diff --git a/tests/data/metadata/fence_ilo.xml b/tests/data/metadata/fence_ilo.xml
index 04f33c8..2219afe 100644
--- a/tests/data/metadata/fence_ilo.xml
+++ b/tests/data/metadata/fence_ilo.xml
@@ -37,7 +37,7 @@
 	<parameter name="notls" unique="0" required="0">
 		<getopt mixed="-t, --notls" />
 		<content type="boolean"  />
-		<shortdesc lang="en">Disable TLS negotiation</shortdesc>
+		<shortdesc lang="en">Disable TLS negotiation, force SSL 3.0</shortdesc>
 	</parameter>
 	<parameter name="passwd" unique="0" required="0">
 		<getopt mixed="-p, --password=[password]" />
@@ -69,6 +69,11 @@
 		<content type="boolean"  />
 		<shortdesc lang="en">SSL connection with verifying fence device's certificate</shortdesc>
 	</parameter>
+	<parameter name="tls1.0" unique="0" required="0">
+		<getopt mixed="--tls1.0" />
+		<content type="boolean"  />
+		<shortdesc lang="en">Disable TLS negotiaton, force TLS 1.0</shortdesc>
+	</parameter>
 	<parameter name="verbose" unique="0" required="0">
 		<getopt mixed="-v, --verbose" />
 		<content type="boolean"  />
diff --git a/tests/data/metadata/fence_ilo2.xml b/tests/data/metadata/fence_ilo2.xml
index 5a090aa..dbc9572 100644
--- a/tests/data/metadata/fence_ilo2.xml
+++ b/tests/data/metadata/fence_ilo2.xml
@@ -37,7 +37,7 @@
 	<parameter name="notls" unique="0" required="0">
 		<getopt mixed="-t, --notls" />
 		<content type="boolean"  />
-		<shortdesc lang="en">Disable TLS negotiation</shortdesc>
+		<shortdesc lang="en">Disable TLS negotiation, force SSL 3.0</shortdesc>
 	</parameter>
 	<parameter name="passwd" unique="0" required="0">
 		<getopt mixed="-p, --password=[password]" />
@@ -69,6 +69,11 @@
 		<content type="boolean"  />
 		<shortdesc lang="en">SSL connection with verifying fence device's certificate</shortdesc>
 	</parameter>
+	<parameter name="tls1.0" unique="0" required="0">
+		<getopt mixed="--tls1.0" />
+		<content type="boolean"  />
+		<shortdesc lang="en">Disable TLS negotiaton, force TLS 1.0</shortdesc>
+	</parameter>
 	<parameter name="verbose" unique="0" required="0">
 		<getopt mixed="-v, --verbose" />
 		<content type="boolean"  />
-- 
1.9.3

